go/vt/wrangler: reduce VReplicationExec calls when getting copy state

maxenglander opened the pull request:
## Description

During `MoveTables SwitchTraffic`, there is a phase where wrangler queries the `copystate` of each stream. Currently it does this by making an individual `VReplicationExec` call for each stream.

This can be prohibitively time-consuming for workflows with very large # of streams. For example, a `MoveTables` workflow where the source and target keyspace have 128 shards, and where the target keyspace has different primary vindex than the source keyspace, will end up with 16384 VStreams. Even if each individual `VReplicationExec` takes only a few milliseconds, in aggregate this could easily take 30+ seconds, risking timing out the `SwitchTraffic` action.

This PR makes things a bit more efficient by making 1 `VReplicationExec` per target shard, and getting all relevant copy states with each call. In my testing with large # of VStreams, this makes the overall `SwitchTraffic` action much faster for the use case I described above. The trade-off here is that the these `VReplicationExec` queries are more expensive with larger result sets.

## Related Issue(s)

Fixes https://github.com/vitessio/vitess/issues/14325

## Checklist

-   [x] "Backport to:" labels are not needed
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on the CI
-   [x] Documentation ~was added or~ is not required

mattlord commented on go/vt/wrangler/vexec.go:
```diff
@@ -739,8 +737,27 @@ func (wr *Wrangler) getStreams(ctx context.Context, workflow, keyspace string) (
+				return nil, err
+			}
+
+			copyStates := copyStatesByVReplID[vreplID]
```
We should check that a value exists, just to be safe, and return an error if not.

maxenglander replied:
Is that definitely an error condition? What if there is a `_vt.vreplication` row but copy state has not been created yet? In this case, `copyStates` will just be `nil`, which seems OK to me.

mattlord commented on go/vt/wrangler/vexec.go:
```diff
@@ -902,26 +919,35 @@ func (wr *Wrangler) printWorkflowList(keyspace string, workflows []string) {
-		id, id)
-	qr, err := wr.VReplicationExec(ctx, tablet.Alias, query)
+func (wr *Wrangler) getCopyStates(ctx context.Context, tablet *topo.TabletInfo, ids []int64) (map[int64][]copyState, error) {
+	var idStrs []string
+	for _, id := range ids {
+		idStrs = append(idStrs, strconv.FormatInt(id, 10))
+	}
+	idsStr := strings.Join(idStrs, ",")
+	cs := make(map[int64][]copyState)
+	query := fmt.Sprintf("select vrepl_id, table_name, lastpk from _vt.copy_state where vrepl_id in (%s) and id in (select max(id) from _vt.copy_state where vrepl_id in (%s) group by vrepl_id, table_name)",
+		idsStr, idsStr)
+	qr, err := wr.tmc.VReplicationExec(ctx, tablet.Tablet, query)
 	if err != nil {
 		return nil, err
 	}
 
 	result := sqltypes.Proto3ToResult(qr)
```
```suggestion
	idsBV, err := sqltypes.BuildBindVariable(ids)
	if err != nil {
		return nil, err
	}
	query, err := sqlparser.ParseAndBind("select vrepl_id, table_name, lastpk from _vt.copy_state where vrepl_id in %a and id in (select max(id) from _vt.copy_state where vrepl_id in %a group by vrepl_id, table_name)",
		idsBV, idsBV)
	if err != nil {
		return nil, err
	}
	qr, err := wr.tmc.VReplicationExec(ctx, tablet.Tablet, query)
	if err != nil {
		return nil, err
	}

	cs := make(map[int64][]copyState)
	result := sqltypes.Proto3ToResult(qr)
```

mattlord reviewed (commented):
Thank you, @maxenglander ! This is a nice little optimization.

I had some minor suggestions. Beyond that, we'll need to make the equivalent optimization for vtctldclient as well. vtctlclient — which is going away soon — uses wrangler whereas vtctldclient uses the workflow server. So we'd make the same optimization here: https://github.com/vitessio/vitess/blob/2f56827b0e6d42b344d43d6e178c351d8595a54e/go/vt/vtctl/workflow/server.go#L1053-L1083

maxenglander commented:
@mattlord I can add more tests if you think it's needed, but I think the code changes are in a decent place. I decided not to try the `JOIN` approach we chatted about in Slack.

mattlord commented:
My concern is that `GetWorkflows` is used quite heavily in vtctldclient (the client going forward). This work makes that significantly heavier for 99.9999% of cases in order to improve the 0.00001% of cases (high number of shards and changing vindexes during the move). It's already relatively heavy and gets called fairly often.

Did you get a sense of how much slower this made the average `workflow show` or `moveables show` commands?

maxenglander commented:
Hey @mattlord the use case we had was a production MySQL cluster with 110 shards that we were migrating into a Vitess keyspace via external tablets into a 128-shard keyspace.

Because the source keyspace had different primary vindexes than the target keyspace, this exploded into 14080 VStreams, and therefore 14080 `getCopyState` calls under the current implementation.

We were seeing that the overall time to complete this [block of code](https://github.com/vitessio/vitess/blob/v16.0.2/go/vt/wrangler/vexec.go#L588) on `SwitchTraffic` take between 40-80s by itself. 

The fact that it was taking so long resulted in various timeouts and short-circuits:

 * remote operation timeout
 * etcd topo lease timeout

...as well as this error:

```
cannot switch traffic for workflow import_workflow at this time: replication lag 61s is higher than allowed lag 30s
```

> This work makes that significantly heavier

Can you break down for me how that is the case? My understanding is that the current implementation fetches copy state once per each stream. I think this PR will change things so that it does the same or else fewer number of calls.

mattlord commented:
> Hey @mattlord the use case we had was a production MySQL cluster with 110 shards that we are migrating from an external keyspace with external tablets into a Vitess keyspace with 128 shards.
> 
> Because the source keyspace has different primary vindexes than the target keyspace, this exploded into 14080 VStreams, and therefore 14080 `getCopyState` calls under the current implementation.

I understand. My point was only that to my knowledge this is an exceedingly rare use case in the history of Vitess. I'm not saying that it's an invalid one. My point was that we should not make things worse/slower for the typical case in order to improve this one. That was my concern. It's a matter of HOW we address that use case/issue, not IF.

> We were seeing that the overall time to complete this [block of code](https://github.com/vitessio/vitess/blob/v16.0.2/go/vt/wrangler/vexec.go#L588) on `SwitchTraffic` was taking between 40-80s by itself.
> 
> The fact that it was taking so long resulted in various timeouts:
> 
> * remote operation timeout
> * etcd topo lease timeout

We may have been overloading various resources like the topo server which has a cascading effect. As we process each result we also make a topo call:
```
si, err := wr.ts.GetShard(ctx, keyspace, primary.Shard)
```
And those results are processed serially. So if the topo responses are a little slow, that will cause the total time to climb a lot in this particular case. We could process those results concurrently as well, synchronizing on the actual updates to the map (but most importantly making those topo calls in parallel).

> ...as well as this error:
> 
> ```
> cannot switch traffic for workflow import_workflow at this time: replication lag 61s is higher than allowed lag 30s
> ```
> 
>  > This work makes that significantly heavier
> 
> Can you break down for me how that is the case? My understanding is that the current implementation fetches copy state once per each stream. I think this PR will change things so that it does the same or else fewer number of calls.

You noted the trade-off yourself in the PR description.
> The trade-off here is that the these VReplicationExec queries are more expensive with larger result sets.

The main query goes from being a point select to a range query. And this can have an impact when subsequently getting the log records for the stream(s) etc as well. I made this more efficient here: https://github.com/vitessio/vitess/pull/14212 It's still a general concern of mine going forward though. So I'm a little (overly) paranoid about it as a lot of vreplication in v18+ is potentially impacted.

All that being said, in general I think we are offsetting any additional cost here by batching things like getting the copy states (although in most cases there will only be one stream on a tablet, but still) so this may be a wash in the end or even improve the typical case. So let me just review in full again. 😄

mattlord commented on go/vt/vtctl/workflow/server.go:
```diff
@@ -1047,16 +1108,24 @@ func (s *Server) getWorkflowState(ctx context.Context, targetKeyspace, workflowN
 	return ts, state, nil
 }
 
-func (s *Server) getWorkflowCopyStates(ctx context.Context, tablet *topo.TabletInfo, id int64) ([]*vtctldatapb.Workflow_Stream_CopyState, error) {
+func (s *Server) getWorkflowCopyStates(ctx context.Context, tablet *topo.TabletInfo, streamIds []int64) ([]*vtctldatapb.Workflow_Stream_CopyState, error) {
 	span, ctx := trace.NewSpan(ctx, "workflow.Server.getWorkflowCopyStates")
 	defer span.Finish()
 
 	span.Annotate("keyspace", tablet.Keyspace)
 	span.Annotate("shard", tablet.Shard)
 	span.Annotate("tablet_alias", tablet.AliasString())
-	span.Annotate("vrepl_id", id)
+	span.Annotate("num_stream_ids", len(streamIds))
```
We should instead annotate it with a CSV list of IDs. The len isn't as interesting IMO and we lost info.

mattlord commented on go/vt/vtctl/workflow/server.go:
```diff
@@ -1069,10 +1138,15 @@ func (s *Server) getWorkflowCopyStates(ctx context.Context, tablet *topo.TabletI
 
 	copyStates := make([]*vtctldatapb.Workflow_Stream_CopyState, len(result.Rows))
 	for i, row := range result.Rows {
-		// These fields are technically varbinary, but this is close enough.
+		streamId, err := row[0].ToCastInt64()
+		if err != nil {
+			return nil, fmt.Errorf("failed to cast vrepl_id to int64: %v", err)
+		}
```
This comment was about the table name and lastpk. There's no reason to case the vrepl_id field: https://github.com/vitessio/vitess/blob/main/go/vt/sidecardb/schema/vreplication/copy_state.sql#L20

mattlord reviewed (commented):
I'm sorry again for the delay. I think this looks OK, but we don't seem to have any test coverage do we? It looks like we updated the existing tests to adjust for the query changes but we don't have any tests that cover the case we're doing the work for do we? Meaning, cases where there are multiple streams per tablet.

Can we add some? Or let me know if I just missed it. I'm talking about unit tests here as we do have some coverage in the endtoend tests as there are cases where there's N streams per tablet (e.g. shard merges).

maxenglander commented:
> Did you get a sense of how much slower this made the average workflow show or moveables show commands?

I realized that I completely misread this initially. I thought you were asking me how much slower the `MoveTables` commands were in the extreme case (14k shards). So my last comment completely missed the mark, sorry.

I did not get a sense of how much slower this was for the average case, although I tested it out locally a bunch with `examples/local` and did not notice any difference.

> Can we add some?

Definitely. I was hoping to get a some initial feedback before investing in tests, which you've now given, and I appreciate! I'll get to work on tests 👷

maxenglander commented on go/vt/wrangler/vexec.go:
```diff
@@ -902,26 +918,37 @@ func (wr *Wrangler) printWorkflowList(keyspace string, workflows []string) {
 	}
 
+	cs := make(map[int64][]copyState)
 	result := sqltypes.Proto3ToResult(qr)
```
`result` can be undefined, so need to rework this a bit. Also `make(map, x, y)` is not valid syntax it seems.

maxenglander commented:
>  It looks like we updated the existing tests to adjust for the query changes but we don't have any tests that cover the case we're doing the work for do we? Meaning, cases where there are multiple streams per tablet.

I took a somewhat lazy approach and just updated the tests to unit tests to have two tables, and test from there.

mattlord commented on go/vt/vtctl/workflow/server.go:
```diff
@@ -419,7 +420,72 @@ func (s *Server) GetWorkflows(ctx context.Context, req *vtctldatapb.GetWorkflows
+	}
+
+	var (
+		fetchCopyStatesWg     sync.WaitGroup
+		fetchCopyStatesErrors concurrency.FirstErrorRecorder
```
I apologize if we discussed this before, but this seems almost tailor made for `sync.errgroup`. Any reason(s) not to leverage that? It can cut the work short when there's an error too (`errgroup.WithContext`) since it's wasted effort after that in this case, and that could be a lot of effort when we're talking about 10s or even 100s of thousands of shard streams.

You can see examples in the files being changed in the PR.

maxenglander replied:
Hm there are no examples in the files changed in this PR of `errgroup` usage.

```sh
$ grep errgroup $(git diff --name-only main | xargs)
$
```

However, I'm familiar with the API and happy to rework this code to use it if you prefer.

mattlord replied:
```
❯ git grep errgroup | grep -E 'wrangler|workflow'
go/vt/vtctl/workflow/traffic_switcher.go:	"golang.org/x/sync/errgroup"
go/vt/vtctl/workflow/traffic_switcher.go:	searchGroup, gctx := errgroup.WithContext(ctx)
go/vt/vtctl/workflow/traffic_switcher.go:	initGroup, gctx := errgroup.WithContext(ctx)
go/vt/wrangler/traffic_switcher.go:	"golang.org/x/sync/errgroup"
go/vt/wrangler/traffic_switcher.go:	searchGroup, gctx := errgroup.WithContext(ctx)
go/vt/wrangler/traffic_switcher.go:	initGroup, gctx := errgroup.WithContext(ctx)
```

mattlord commented on go/vt/vtctl/workflow/server.go:
```diff
@@ -1053,16 +1115,28 @@ func (s *Server) getWorkflowState(ctx context.Context, targetKeyspace, workflowN
 	span.Annotate("shard", tablet.Shard)
 	span.Annotate("tablet_alias", tablet.AliasString())
-	span.Annotate("vrepl_id", id)
+	var strStreamIds []string
+	for _, streamId := range streamIds {
+		strStreamIds = append(strStreamIds, strconv.FormatInt(streamId, 10))
+	}
+	span.Annotate("stream_ids", strings.Join(strStreamIds, ","))
```
IMO it would adequate in this case to do this since there can't be any whitespace within the values:
```
span.Annotate("stream_ids", fmt.Sprintf("%v", strStreamIds))
```

Or this (https://go.dev/play/p/VdgdivLDiA8):
```
span.Annotate("stream_ids", fmt.Sprintf("%#v", strStreamIds))
```

mattlord commented on go/vt/vtctl/workflow/server.go:
```diff
@@ -1075,10 +1149,15 @@ func (s *Server) getWorkflowCopyStates(ctx context.Context, tablet *topo.TabletI
-		// These fields are technically varbinary, but this is close enough.
+		streamId, err := row[0].ToInt64()
+		if err != nil {
+			return nil, fmt.Errorf("failed to cast vrepl_id to int64: %v", err)
```
IMO it's better to use vterrors in new code so that we can rely on having a meaningful code in more places over time. It's also helpful in these rare cases if we print the value we couldn't cast using `row[0].ToString()`. For example:
```
return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "failed to cast vrepl_id value of %q to int64: %v", row[0].ToString(), err)
```

mattlord commented on go/vt/wrangler/vexec.go:
```diff
@@ -736,8 +733,27 @@ func (wr *Wrangler) getStreams(ctx context.Context, workflow, keyspace string) (
 			continue
 		}
+		// Get all copy states for the shard.
+		var vreplIDs []int64
```
We should allocate space once:
```
vreplIDs := make([]int64, 0, len(nqr.Rows))
```

mattlord reviewed (commented):
This looks good to me. Thanks, @maxenglander ! I only had some minor nits and suggestions. We can discuss/address those along with any that @rohit-nayak-ps may have. @rohit-nayak-ps can you please also review this whenever you have time?

mattlord reviewed (approved):
Thanks again, @maxenglander !
